File: leetcode/tree/257. Binary Tree Paths/py/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        def dfs(node, path):
            if node is None:
                return

            val = str(node.val)

            if not node.left and not node.right:
                suffix = '->' + val if len(path) else val
                self.ans.append('->'.join(path) + suffix)
                return

            path.append(val)
            dfs(node.left, path)
            path.pop()

            path.append(val)
            dfs(node.right, path)
            path.pop()

        self.ans = []

        dfs(root, [])

        return self.ans

File: leetcode/tree/257. Binary Tree Paths/js/main.js
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
const binaryTreePaths = function (root) {
  let path = [];
  let paths = [];

  const treePaths = (node, path, paths) => {
    if (node === null) return;

    path.push(node.val);

    if (node.left === null && node.right === null) {
      let str = path.join('->');
      paths.push(str);
    }

    treePaths(node.left, path, paths);
    treePaths(node.right, path, paths);
    path.pop();
  };

  treePaths(root, path, paths);

  return paths;
};
